qa/L0_logging/log_format_test.py

```diff
@@ -0,0 +1,470 @@
+    hours = int(hours)
+    minutes = int(minutes)
+    seconds = float(seconds)
+    return datetime.timedelta(hours=hours, minutes=minutes, seconds=seconds)
+
+
+validators = {}
+
+
+def validator(func):
+    validators[func.__name__.replace("validate_", "")] = func
+    return func
+
+
+@validator
+def validate_level(level, _):
+    assert level in LEVELS
+
+
+@validator
+def validate_month(month, _):
+    assert month.isdigit()
+    month = int(month)
+    assert month >= 1 and month <= 12
+
+
+@validator
+def validate_day(day, _):
+    assert day.isdigit()
+    day = int(day)
+    assert day >= 1 and day <= 31
+
+
+@validator
+def validate_ISO8601_timestamp(timestamp, _):
+    datetime.datetime.fromisoformat(timestamp.rstrip("Z"))
+
+
+@validator
+def validate_timestamp(timestamp, _):
+    parse_timestamp(timestamp)
+
+
+@validator
+def validate_pid(pid, _):
+    assert pid.isdigit()
+
+
+@validator
+def validate_file(file_, _):
+    assert Path(file_).name is not None
+
+
+@validator
+def validate_line(line, _):
+    assert line.isdigit()
+
+
+def split_row(row):
+    return [r.strip() for r in row.group("row").strip().split("|")]
+
+
+def validate_protobuf(protobuf):
+    # Note currently we only check for model config
+    # but technically any protubuf should be valid
+
+    google.protobuf.text_format.ParseLines(
+        protobuf, grpcclient.model_config_pb2.ModelConfig()
+    )
+
+
+def validate_table(table_rows):
+    index = 0
+    top_border = table_border_regex.search(table_rows[index])
+    assert top_border
+
+    index += 1
+    header = table_row_regex.search(table_rows[index])
+    assert header
+    header = split_row(header)
+
+    index += 1
+    middle_border = table_border_regex.search(table_rows[index])
+    assert middle_border
+
+    # Process each row
+    index += 1
+    parsed_rows = []
+    row = ""
+    for index, row in enumerate(table_rows[index:]):
+        matched = table_row_regex.search(row)
+        if matched:
+            row_data = split_row(matched)
+            parsed_rows.append(row_data)
+
+    end_border = table_border_regex.search(row)
+    assert end_border
+
+    for row in parsed_rows:
+        assert len(row) == len(header)
+
+
+@validator
+def validate_message(message, escaped):
+    split_message = message.split("\n")
+    heading = split_message[0]
+    obj = split_message[1:] if len(split_message) > 1 else []
+    if heading and escaped:
+        try:
+            json.loads(heading)
+        except Exception as e:
+            raise Exception(
+                f"{e.__class__.__name__} {e}\nFirst line of message in log record is not a valid JSON string"
+            )
+    elif heading:
+        with pytest.raises(json.JSONDecodeError):
+            json.loads(heading)
+    if obj:
+        match = table_border_regex.search(obj[0])
+        if match:
+            validate_table(obj)
+        elif escaped:
+            validate_protobuf(obj)
```

Comment:
Deferred to follow-up PR, but would be nice to have some comments explaining the validations that the only acceptable multi-line logs will be tables/protobuf (perhaps json?), but will otherwise be restricted to single-line logs if escaped correctly.